services/auth/oauth2.go

```diff
@@ -58,15 +173,176 @@ func (o *OAuth2) Name() string {
+		kid, ok := tokenSHA.Header["kid"].(string)
+		if !ok {
+			log.Error("Can't Get KID .")
+			return nil, fmt.Errorf("kid header not found")
+		}
+		return GetPublicKey(kid)
+	})
+	// check if the any error when they parse the access token
+	if err != nil || !token.Valid {
+		log.Error("Error When Get Public Key %s ", err)
+		return nil, fmt.Errorf("error when get public key %s ", err)
+	}
+	// EXTRACT DATA FROM ACCESS TOKEN
+	claims, ok := token.Claims.(*accessToken)
+	if !ok {
+		// CAN'T PARSE CLAIMS
+		return nil, fmt.Errorf("can't parse claims")
+
+	}
+
+	// CHECK ON THE EXPIRE DATE
+	if claims.ExpiresAt < time.Now().UTC().Unix() {
+		// ACCESS TOKEN IS EXPIRE !
+		return nil, fmt.Errorf("access token is expired ")
+
+	}
+
+	// CHECK IF THE USER ALREADY EXIST
+	u, err := user_model.GetUserByName(db.DefaultContext, claims.Username)
+
+	// IF THE USER NOT EXISTS
+	if err != nil {
+		provider, err := setting.GetAWSKey(setting.SecAWS, setting.KeyClient)
+		if err != nil {
+			return nil, err
+		}
+		sess := session.Must(session.NewSession())
+		reg := strings.Split(claims.Iss, ".")[1]
+		sess.Config.Region = aws.String(reg)
+		cognitoClient := cognitoidentityprovider.New(sess)
+		res, err := cognitoClient.GetUser(&cognitoidentityprovider.GetUserInput{AccessToken: &tokenSHA})
+		if err != nil {
+			return nil, fmt.Errorf("ACCESS TOKEN NOT VALID ON COGNITO")
+
+		}
+
+		var emailIndex int
+		for i := 0; i < len(res.UserAttributes); i++ {
+			if *res.UserAttributes[i].Name == "email" {
+				emailIndex = i
+			}
+		}
+
+		loginSource, err := auth_model.GetActiveOAuth2SourceByName(provider.Value())
+		if err != nil {
+			log.Debug("%v", err)
+			return nil, err
+		}
+		if loginSource == nil {
+
+			if err := auth_model.CreateSource(&auth_model.Source{
+				Type:          auth_model.OAuth2,
+				Name:          provider.Value(),
+				IsActive:      true,
+				IsSyncEnabled: true,
+				Cfg:           &oauth2.Source{},
+			}); err != nil {
+				if models.IsErrLoginSourceAlreadyExist(err) {
+					return nil, err.(models.ErrLoginSourceAlreadyExist)
+				} else {
+					return nil, err
+				}
+			}
+			loginSource, err = auth_model.GetActiveOAuth2SourceByName(provider.Value())
+			if err != nil || loginSource == nil {
+				return nil, err
+			}
+		}
+
+		log.Debug("User will be created")
+		// CREATE NEW USER
+		u = &user_model.User{
+			LowerName:   strings.ToLower(*res.Username),
+			Name:        *res.Username,
+			LoginType:   auth_model.OAuth2,
+			LoginSource: loginSource.ID,
+			LoginName:   *res.Username,
+			Email:       *res.UserAttributes[emailIndex].Value,
+			IsActive:    !(setting.Service.RegisterEmailConfirm || setting.Service.RegisterManualConfirm),
+		}
+
+		err = user_model.CreateUser(u)
+		if err != nil {
+			fmt.Println(err)
+			// SOME ERROR ACCURE WHEN CREATING USER
+			return nil, fmt.Errorf("CAN'T CREATE NEW USER WITH COGNITO ACCES TOKEN %s ", err)
+		}
+		// Auto-set admin for the only user.
+		if user_model.CountUsers(nil) == 1 {
```

Comment:
In this case we make first user in DB is with admin role but we need to recheck on this case with eng @maha-saad